renku/core/metadata/repository.py

```diff
@@ -0,0 +1,1192 @@
+
+    @property
+    def modified_changes(self) -> List["Diff"]:
+        """Return a list of changes that are not staged."""
+        try:
+            diff = self._repository.index.diff(None, ignore_submodules=True)
+            return [
+                Diff(
+                    a_path=git_unicode_unescape(d.a_path),
+                    b_path=git_unicode_unescape(d.b_path),
+                    change_type=d.change_type,
+                )
+                for d in diff
+            ]
+        except (git.BadName, git.BadObject, git.GitError) as e:
+            raise errors.GitError("Cannot get modified changes") from e
+
+    @property
+    def unmerged_blobs(self) -> Dict[str, List[Tuple[int, "Object"]]]:
+        """Return a map of path to stage and blob for unmerged blobs in the current index."""
+        return {
+            str(path): [(e[0], Object.from_object(e[1])) for e in values]
+            for path, values in self._repository.index.unmerged_blobs().items()
+        }
+
+    @property
+    def untracked_files(self) -> List[str]:
+        """Return the list of untracked files."""
+        return self._repository.untracked_files
+
+    @property
+    def files(self) -> List[str]:
+        """Return a list of all files in the current version of the repository."""
+        return [e[0] for e in self._repository.index.entries]
+
+    def is_valid(self) -> bool:
+        """Return True if a valid repository exists."""
+        return self._repository is not None
+
+    def add(self, *paths: Union[Path, str], force: bool = False, all: bool = False) -> None:
+        """Add a list of files to be committed to the VCS."""
+        if all:
+            assert len(paths) == 0, "Cannot pass both ``all`` and ``paths``."
+            self.run_git_command("add", all=True, force=force)
+        else:
+            for batch in split_paths(*paths):
+                self.run_git_command("add", *batch, force=force)
+
+    def commit(
+        self,
+        message: str,
+        *,
+        amend: bool = False,
+        author: "Actor" = None,
+        committer: "Actor" = None,
+        no_verify: bool = False,
+        no_edit: bool = False,
+    ) -> "Commit":
+        """Commit added files to the VCS."""
+        env = {}
+        if author:
+            env.update({"GIT_AUTHOR_NAME": author.name, "GIT_AUTHOR_EMAIL": author.email})
+        if committer:
+            env.update({"GIT_COMMITTER_NAME": committer.name, "GIT_COMMITTER_EMAIL": committer.email})
+
+        self.run_git_command("commit", message=message, no_verify=no_verify, amend=amend, no_edit=no_edit, env=env)
+
+        return Commit.from_commit(self._repository, self._repository.head.commit)
+
+    def checkout(self, reference: Union["Branch", str]):
+        """Check-out a specific reference."""
+        self.run_git_command("checkout", reference)
+
+    def clean(self):
+        """Remove all untracked files."""
+        self.run_git_command("clean", "-xdff")
```

Comment:
Is the double `f` here a typo? I don't see anything in the git clean command that would suggest it does something special

Reply:
Haha, I had to double check too! The doc says: `Git will refuse to modify untracked nested git repositories (directories with a .git subdirectory) unless a second -f is given.` (`git version 2.24.3 (Apple Git-128)`).